pkg/server/admin.go

```diff
@@ -1040,14 +1041,44 @@ func (s *adminServer) Cluster(
 func (s *adminServer) Health(
 	ctx context.Context, req *serverpb.HealthRequest,
 ) (*serverpb.HealthResponse, error) {
+	isLive, err := s.server.nodeLiveness.IsLive(s.server.NodeID())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "node is not live")
+	}
+	if !isLive {
+		return nil, status.Errorf(codes.Unavailable, "node is not live")
+	}
+	return &serverpb.HealthResponse{}, nil
+}
+
+// Ready returns the target node's readiness to service client requests.
+func (s *adminServer) Ready(
+	ctx context.Context, req *serverpb.ReadyRequest,
+) (*serverpb.ReadyResponse, error) {
 	isHealthy, err := s.server.nodeLiveness.IsHealthy(s.server.NodeID())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, err.Error())
 	}
 	if !isHealthy {
-		return nil, status.Errorf(codes.Unavailable, "node is not healthy")
+		return nil, status.Errorf(codes.Unavailable, "node is not ready")
 	}
-	return &serverpb.HealthResponse{}, nil
+
+	// Attempt to execute a SQL query with a timeout.
```

Comment:
Can you explain for posterity which situations this will catch?

Reply:
I changed this to observe the `serveMode` on the server, which is set after startup. I think this is the best way to check for readiness since more involved checks are not necessarily what we're going for here. i.e. we're not trying to answer whether or not we can execute SQL queries at the time of the check, we're trying to answer if the server should be able to accept connections.